### PR TITLE
Fixed bug that causes the server to run out of peers eventually. 

### PR DESCRIPTION
**Pull request by DanielKrawisz**
This pr comprises a number of changes that resulted from my investigations into bmd's behavioral problems.

* Fixed bug that causes the server to run out of peers eventually. The problem was caused a map called outboundGroups which tracked the kinds of ip addresses that the server was connected to. This map was updated correctly when a peer joined but not when it disconnected. Consequently, the server eventually reached a state where it could not connect to any address because every one was already registered in outboundGroups. 
* Fixed bug that caused the object manager not to receive the message that a peer is ready to download more. There was on oversight in one of my previous prs in which I forgot to uncomment some lines in peer.go that had been written in anticipation of the object manager being able to download from many peers at once. As a result, the peers do not signal when they are ready to start downloading again. 
* Added comments to addrmgr.updateAddress because that function is extremely confusing. 
* Added a channel to server that that is used to send a message to disconnect a peer. The object manager needs to do this because it can lock up if it tries to disconnect a peer from its own goroutine. 
* There is a problem in which the objectmanager sometimes disconnects several peers at once for expired requests. I think this problem has to do with channel buffers that are too large and which can end up holding messages for a very long time. In general, if we can get by with shorter channel buffers, we should. My previous pr allowed the object manager to function with an unbuffered channel. This one reduces the buffer size on the sendQueue from 20 to 5. I'm not sure the issue has been entirely resolved but it seems to be improved a lot. More experimentation will have to be done to resolve this problem completely.

### CODE REVIEW

File: peer/send.go [hunk 1]
```diff
@@ -21,9 +21,9 @@ import (
 )
 
 const (
-	// queueSize is the size of data and message queues. TODO arbitrarily set as
-	// 20.
-	queueSize = 20
+	// queueSize is the size of data and message queues. The number should be
```
Comment (ishbir) [inline peer/send.go]: s/queueSize/SendQueueSize

File: peer/connection.go [hunk 2]
```diff
@@ -107,8 +109,6 @@ func (pc *connection) ReadMessage() (wire.Message, error) {
 		return nil, err
 	}
 
-	pc.idleTimer.Reset(pc.idleTimeout)
-
```
Comment (ishbir) [inline peer/connection.go]: Why did you remove it from here? I think it should be there in both WriteMessage and ReadMessage.

File: server.go [hunk 3]
```diff
@@ -445,12 +397,13 @@ func (s *server) peerHandler() {
 		case p := <-s.banPeers:
 			s.handleBanPeerMsg(p)
 
+		//
+		case p := <-s.disconPeers:
```
Comment (ishbir) [inline server.go]: Empty comment. Also, why did you decide to move disconnection from peer to server?

File: peer_test.go [hunk 4]
```diff
@@ -1017,6 +1026,7 @@ func TestOutboundPeerHandshake(t *testing.T) {
 
 	cfg.ConnectPeers = []string{}
 	NewConn = peer.NewConnection
+	resetCfg(cfg)()
```
Comment (ishbir) [inline peer_test.go]: Why did you move all the resetCfg calls outside? I had intended that the function would renew config after each test.

File: peer/connection.go (hunk 2 above)
Comment (DanielKrawisz) [inline peer/connection.go]: The purpose is to prevent the remote peer's connection from timing out and to send a pong if it would. We only need to do that for outgoing messages.

File: peer_test.go (hunk 4 above)
Comment (DanielKrawisz) [inline peer_test.go]: One of us doesn't know how to use defer correctly! You put lines that said defer resetCfg(cfg)() in the loop. My understanding is that defer is not called until the function closes, so you would have just queued several calls to resetCfg that all run after all tests are complete.

Comment (ishbir) [inline peer_test.go]: Yeah, the function associated with defer isn't called. But take note that resetCfg returns a function which is then supposed to be called by defer. Therefore, resetCfg is called as soon as it hits the line with defer. But the function it returns (which just deletes the temporary directory), isn't called until the end of the current function.

File: peer/connection.go (hunk 2 above)
Comment (ishbir) [inline peer/connection.go]: But if we received a message from a peer, it means that the peer is active and by definition not idle. Therefore, idleTimer should be reset. Why should it be restricted to outgoing messages?

File: server.go (hunk 3 above)
Comment (DanielKrawisz) [inline server.go]: There is a problem when the object manager tries to disconnect a peer. This tended to flood the object manager's message channel. I created this channel in the server in case there are similar problems in the future. Now the object manager sends this message to the server instead of calling the peer's disconnect method.

File: peer_test.go (hunk 4 above)
Comment (DanielKrawisz) [inline peer_test.go]: Wow, that's crazy! I love it!!! Haha sorry I guess I didn't understand it before. :)

File: peer/connection.go (hunk 2 above)
Comment (DanielKrawisz) [inline peer/connection.go]: You're right. I forgot that tcp included message acknowledgements!

File: peer.go [hunk 5]
```diff
@@ -183,7 +184,15 @@ func (p *bmpeer) PushGetDataMsg(ivl []*wire.InvVect) {
...
-	peerLog.Debug(p.peer.PrependAddr(fmt.Sprint(len(ivl), " requests assigned for a total of ", p.inventory.NumRequests(), ".")))
+	// This can be done once the object manager downloads more intelligently.
+	if p.inventory.NumRequests() > MaxPeerRequests/2 {
+		p.signalReady = p.inventory.NumRequests() / 2
+	} else {
+		p.signalReady = 0
+	}
+	peerLog.Debug(p.peer.PrependAddr(fmt.Sprint(len(ivl),
+		" requests assigned for a total of ", p.inventory.NumRequests(),
+		"; signal ready at ", p.signalReady)))
```
Comment (ishbir) [inline peer.go]: I think MaxPeerRequests should be renamed to maxPeerRequests, like all other constants. Also, the comment on line 187 didn't make sense to me. Could you rewrite or clarify?

File: peer/send.go [hunk 6]
```diff
@@ -21,9 +21,9 @@ import (
 )
 
 const (
-	// queueSize is the size of data and message queues. TODO arbitrarily set as
-	// 20.
-	queueSize = 20
+	// SendQueueSize is the size of data and message queues. The number should be
+	// small if possible, and ideally we would use an unbuffered channel eventually.
+	SendQueueSize = 5
```
Comment (ishbir) [inline peer/send.go]: Constant should be unexported.

File: peer_test.go [hunk 7]
```diff
@@ -1217,6 +1222,8 @@ func TestProcessAddr(t *testing.T) {
 		serv.Stop()
 
 		serv.WaitForShutdown()
+
+		resetCfg(cfg)()
```
Comment (ishbir) [inline peer_test.go]: Remove.

File: peer_test.go (hunk 4 above)
Comment (ishbir) [inline peer_test.go]: Great. :) Then could you remove it from here and put it back where it was? Thanks.

File: server.go [hunk 8]
```diff
@@ -276,56 +278,6 @@ func (s *server) AddNewPeer(addr string, stream uint32, permanent bool) error {
...
-	case getAddedNodesMsg:
-		// Respond with a slice of the relavent peers.
-		peers := make([]*bmpeer, 0, len(s.state.persistentPeers))
-		for p := range s.state.persistentPeers {
-			peers = append(peers, p)
-		}
-		msg.reply <- peers
-	}
-}
-
```
Comment (ishbir) [inline server.go]: Since you removed this function, could you also remove all other functions that use the s.query channel? And also the channel itself.

Comment (DanielKrawisz) [inline server.go]: Hm, what a good idea!

Comment (DanielKrawisz) [inline server.go]: I decided to leave it in because we might want to use it eventually, but we should remember that this is here and might have to be adapted to our needs later.

File: peer.go (hunk 5 above)
Comment (DanielKrawisz) [inline peer.go]: Ok, I changed the name to maxPeerRequests. 

The comment was about something that wasn't there anymore, so I removed it.

File: server.go (hunk 8 above)
Comment (ishbir) [inline server.go]: For reference: As per our discussion, you agreed to removed them.

Comment (DanielKrawisz) [inline server.go]: Indeed.

File: peer_test.go [hunk 9]
```diff
@@ -1015,7 +1020,6 @@ func TestOutboundPeerHandshake(t *testing.T) {
 		serv.WaitForShutdown()
 	}
 
-	cfg.ConnectPeers = []string{}
```
Comment (ishbir) [inline peer_test.go]: Removing this is causing tests to fail. It's needed because resetCfg doesn't reset cfg.ConnectPeers.
